p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,790 @@
+			if parsedAddr.useHTTPS && parsedAddr.sni != "" && parsedAddr.sni != host {
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/tls/sni/%s/http", host, port, parsedAddr.sni))
+			} else {
+				scheme := "http"
+				if parsedAddr.useHTTPS {
+					scheme = "https"
+				}
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/%s", host, port, scheme))
+
+			}
+
+			if parsedAddr.useHTTPS {
+				go func() {
+					srv := http.Server{
+						Handler:   h.ServeMux,
+						TLSConfig: h.TLSConfig,
+					}
+					errCh <- srv.ServeTLS(l, "", "")
+				}()
+				h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
+			} else if h.InsecureAllowHTTP {
+				go func() {
+					errCh <- http.Serve(l, h.ServeMux)
+				}()
+				h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
+			} else {
+				// We are not serving insecure HTTP
+				log.Warnf("Not serving insecure HTTP on %s. Prefer an HTTPS endpoint.", listenAddr)
+			}
+		}
+		return nil
+	}()
+	if err != nil {
+		closeAllListeners()
+		return err
+	}
+
+	close(h.httpTransport.waitingForListeners)
+	closedWaitingForListeners = true
+
+	if len(h.httpTransport.listeners) == 0 || len(h.httpTransport.listenAddrs) == 0 {
+		closeAllListeners()
+		return ErrNoListeners
+	}
+
+	expectedErrCount := len(h.httpTransport.listeners)
+	select {
+	case <-h.httpTransport.closeListeners:
+	case err = <-errCh:
+		expectedErrCount--
+	}
+
+	// Close all listeners
+	closeAllListeners()
+	for i := 0; i < expectedErrCount; i++ {
+		<-errCh
+	}
+	close(errCh)
+
+	return err
+}
+
+func (h *Host) Close() error {
+	h.httpTransportInit()
+	close(h.httpTransport.closeListeners)
+	return nil
+}
+
+// SetHTTPHandler sets the HTTP handler for a given protocol. Automatically
+// manages the .well-known/libp2p mapping.
+// http.StripPrefix is called on the handler, so the handler will be unaware of
+// its prefix path.
+func (h *Host) SetHTTPHandler(p protocol.ID, handler http.Handler) {
+	h.SetHTTPHandlerAtPath(p, string(p), handler)
+}
+
+// SetHTTPHandlerAtPath sets the HTTP handler for a given protocol using the
+// given path. Automatically manages the .well-known/libp2p mapping.
+// http.StripPrefix is called on the handler, so the handler will be unaware of
+// its prefix path.
+func (h *Host) SetHTTPHandlerAtPath(p protocol.ID, path string, handler http.Handler) {
+	if path == "" || path[len(path)-1] != '/' {
+		// We are nesting this handler under this path, so it should end with a slash.
+		path += "/"
+	}
+	h.WellKnownHandler.AddProtocolMeta(p, ProtocolMeta{Path: path})
+	h.serveMuxInit()
+	h.ServeMux.Handle(path, http.StripPrefix(path, handler))
+}
+
+// PeerMetadataGetter lets RoundTrippers implement a specific way of caching a peer's protocol mapping.
+type PeerMetadataGetter interface {
+	GetPeerMetadata() (PeerMeta, error)
+}
+
+type streamRoundTripper struct {
+	server   peer.ID
+	h        host.Host
+	httpHost *Host
+}
+
+type streamReadCloser struct {
```

Comment:
Can you add a comment why you're introducing this type (as opposed to just using the underlying `network.Stream`).